File: 0277-find-the-celebrity/0277-find-the-celebrity.kt
/* The knows API is defined in the parent class Relation.
      fun knows(a: Int, b: Int) : Boolean {} */

class Solution: Relation() {
    override fun findCelebrity(n: Int) : Int {
        var cand = 0
        for(i in 1..<n) {
            if(knows(cand, i)) {
                cand = i
            }
        }
        for(i in 0..<n) {
            if(cand == i) {
                continue
            }
            if(knows(cand, i) or !knows(i, cand)) {
                return -1
            }
        }
        return cand
	}
}

File: 0277-find-the-celebrity/0277-find-the-celebrity.cpp
/*
Took: 20m 10s
Time: O(n)
Space: O(1)
*/


/* The knows API is defined for you.
      bool knows(int a, int b); */

class Solution {
public:
    int findCelebrity(int n) {
        int cand = 0;
        for(int i = 1; i < n; ++i) {
            if(knows(cand, i)) {
                cand = i;
            }
        }
        for(int i = 0; i < n; ++i) {
            if(i == cand) {
                continue;
            }
            if(knows(cand, i) || !knows(i, cand)) {
                return -1;
            }
        }
        return cand;
    }
};